src/main/java/com/nextcloud/common/OkHttpMethodBase.kt

```diff
@@ -56,32 +56,53 @@ abstract class OkHttpMethodBase(var uri: String,
         return httpBuilder.build()
     }
 
+    /**
+     * Set request headers completely replacing existing headers.
+     * To clear request headers, call this method with empty list.
+     *
+     * @param headers List of header-value pairs
+     */
+    fun setRequestHeaders(vararg headers: Pair<String, String>) {
+        requestHeaders.clear()
+        requestHeaders.putAll(headers)
+    }
+
+    /**
+     * Adds request header, overwriting any existing value.
+     *
+     * @param header HTTP request header name
+     * @param value HTTP request header value
+     */
+    fun addRequestHeader(header: String, value: String) {
+        requestHeaders.put(header, value)
+    }
+
     fun setQueryString(params: Map<String, String>) {
         queryMap = params
     }
 
     fun getResponseBodyAsString(): String {
-        return response.body()?.string() ?: ""
+        return response?.body()?.string() ?: ""
     }
 
     fun releaseConnection() {
-        response.body()?.close()
+        response?.body()?.close()
     }
 
     fun getStatusCode(): Int {
-        return response.code()
+        return response?.code() ?: 0
```

Comment:
0 as status code for failing response seems to be wrong, or?
Maybe 400?

Reply:
400 seems to be the best guess while I can't say if is is always a client error and if bad request is always right, but as a fallback, I'd say it is fine.

Reply:
The whole problem whith the API is that it tries to get some data from non-existing entity.
There is no response object so there is no status code as well and there is also no good answer what to do with that.
0 seems to be a good choice for a 'non-value'. Apache is using -1. Good as well.

400 is not universally valid proposition - what is resposne is null because theere were not network? In such case, no http request was ever made, so there is no code, headers, body, etc.
Whatever we do, 0 or -1 is still better than NPE, IMO.

Reply:
Yeah, 400 would indicate that at least something was tried, but it is more like an uninitialized state, and thus -1 makes more sense.